File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
def return_map(lst, add_time):
    return [{'time': v['time'] + add_time, 'w': v['w']} for v in lst]

def solution(bridge_length, bridge_weight, truck_weights):
    on_bridge = []
    on_weight = 0
    second = 0

    def on_to_bridge(current_weight):
        nonlocal on_weight, second, on_bridge
        on_bridge.append({'time': 0, 'w': current_weight})
        on_weight += current_weight
        second += 1
        on_bridge = return_map(on_bridge, 1)

    def down_from_bridge():
        nonlocal on_weight, on_bridge
        while on_bridge and on_bridge[0]['time'] >= bridge_length:
            down = on_bridge.pop(0)
            on_weight -= down['w']

    def wait_for_down():
        nonlocal second, on_bridge
        wait = bridge_length - on_bridge[0]['time']
        on_bridge = return_map(on_bridge, wait)
        down_from_bridge()
        second += wait

    for w in truck_weights:
        if w + on_weight <= bridge_weight:
            on_to_bridge(w)
        else:
            if on_bridge[0]['time'] > bridge_length:
                down_from_bridge()
            else:
                wait_for_down()
            while w + on_weight > bridge_weight:
                wait_for_down()
            on_to_bridge(w)

    while on_bridge:
        wait_for_down()

    return second + 1

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Queue<int[]> bridge = new LinkedList<>();
        int time = 0; 
        int onWeight = 0; 
        int idx = 0; // 대기 트럭 인덱스
        
        while (idx < truck_weights.length || !bridge.isEmpty()) {
            time++; 
            
            if (!bridge.isEmpty() && bridge.peek()[1] == time) {
                onWeight -= bridge.poll()[0];
            }
            
            // 대기 트럭 중 다음 트럭이 다리로 올라갈 수 있는지 확인
            if (idx < truck_weights.length && onWeight + truck_weights[idx] <= weight && bridge.size() < bridge_length) {
                // [1]에 그냥 언제 나갈 수 있는 지를 집어넣음.
                bridge.add(new int[]{truck_weights[idx], time + bridge_length}); 
                onWeight += truck_weights[idx]; // 다리 위에 올라간 트럭의 무게를 더함
                idx++; // 다음 대기 트럭으로 이동
            }
        }
        
        return time;
    }
}
